### PR TITLE
Linter refactor

### PR DESCRIPTION
**Pull request by glasss13**
I got rid of the timeout on the python spawner side and kept it only on the spawner side. Additionally, I made the python spawner send its status via pipe to the spawner and then the spawner is responsible for setting the appropriate firebase status, this avoids having both of them having to coordinate through firebase.

### CODE REVIEW

**Comment by glasss13** [comment]
The spawner no longer requires any network access as the linter server now sends the algorithm to the linter spawner through a pipe.

**Comment by egelja** [comment]
will review tomorrow

File: linter/src/crow/crow.cpp [hunk 1]
```diff
@@ -43,83 +44,44 @@ get_server_thread()
...
+                        uid
+                    )
+                );
+                crow::json::wvalue response = crow::json::wvalue({
+                    {"linting_status",
+                     static_cast<int>(client::LintingResultOption::UNKNOWN)}
+                });
+
+                res.body = response.dump();
+                res.code = 200;
```
Comment (egelja) [inline linter/src/crow/crow.cpp:64]: wouldn't this be 404 or 400?

File: linter/src/crow/crow.cpp [hunk 2]
```diff
@@ -43,83 +44,44 @@ get_server_thread()
...
-                            nutc::client::set_lint_failure(uid, algo_id, error_msg);
+            client::LintingResultOption algo_status_code;
+            auto lint_res = spawner_manager.spawn_client(algo_code.value());
+            std::cout << lint_res.message << '\n';
+            if (lint_res.success) {
+                nutc::client::set_lint_success(uid, algo_id, lint_res.message);
+                algo_status_code = client::LintingResultOption::SUCCESS;
+            }
+            else {
+                nutc::client::set_lint_failure(uid, algo_id, lint_res.message);
```
Comment (egelja) [inline linter/src/crow/crow.cpp:77]: these web requests could (and probably should) be done asynchronously.

File: linter/src/firebase/fetching.cpp [hunk 3]
```diff
@@ -127,18 +122,10 @@ get_algo(const std::string& uid, const std::string& algo_id)
     glz::json_t user_info = get_user_info(uid);
     // if not has "algos"
     if (!user_info.contains("algos")) {
-        log_w(
-            firebase, "User {} has no algos. Will not participate in simulation.", uid
-        );
-        return std::nullopt;
-    }
-    if (!user_info["algos"].contains(algo_id)) {
```
Comment (egelja) [inline linter/src/firebase/fetching.cpp:135]: why did you remove this check?

File: linter/src/firebase/fetching.cpp [hunk 4]
```diff
@@ -212,17 +189,13 @@ firebase_request(
...
 
         res = curl_easy_perform(curl);
-        if (res != CURLE_OK) {
-            log_e(firebase, "curl_easy_perform() failed: {}", curl_easy_strerror(res));
-        }
 
         curl_easy_cleanup(curl);
     }
     glz::json_t json{};
     auto error = glz::read_json(json, readBuffer);
```
Comment (egelja) [inline linter/src/firebase/fetching.cpp:196]: yyou should probably check `res` here on L222 before bothering to read the JSON.

File: linter/src/mock_api/mock_api.cpp [hunk 5]
```diff
@@ -2,25 +2,11 @@
...
-            "Mock API: Placing order side {} ticker {} quantity {} price "
-            "{}",
-            side,
-            ticker,
-            quantity,
-            price
-        );
-        return true;
-    };
+    return [](const std::string&, const std::string&, float, float) { return true; };
```
Comment (egelja) [inline linter/src/mock_api/mock_api.cpp:9]: why does this function return a closure that doesn't capture anything???

delete these two files
and add this to `lint.cpp`:

```cpp
namespace {

bool 
mock_market_func(const std::string&, const std::string&, float, float)
{
    return true;
}

} // namespace

File: linter/src/lint_result.hpp
Comment (egelja) [inline linter/src/lint_result.hpp:1]: move this to the `lint` subdir

File: linter/src/lint/lint.cpp [hunk 6]
```diff
@@ -1,58 +1,52 @@
...
-                      << " for uid " << uid << "\n";
-        nutc::client::set_lint_failure(uid, algo_id, "Could not find algo id");
-        return false;
-    }
 
+lint_result
+lint(const std::string& algo_code)
+{
+    std::string out_message = "[linter] starting to lint algorithm\n";
     bool e = nutc::pywrapper::create_api_module(nutc::mock_api::getMarketFunc());
```
Comment (egelja) [inline linter/src/lint/lint.cpp:19]:
```suggestion
    bool ok = nutc::pywrapper::create_api_module(nutc::mock_api::getMarketFunc());
```

File: linter/src/mock_api/mock_api.cpp (hunk 5 above)
Comment (egelja) [reply linter/src/mock_api/mock_api.cpp:9]: then you can call `create_api_module(mock_market_func)`.

File: linter/src/pywrapper/runtime.cpp [hunk 7]
```diff
@@ -28,10 +31,30 @@ create_api_module(
     return true;
 }
 
+bool
+supress_stdout()
+{
+    try {
+        py::exec(R"(
```
Comment (egelja) [inline linter/src/pywrapper/runtime.cpp:38]: extract the string to a const global.

File: linter/src/pywrapper/runtime.cpp [hunk 8]
```diff
@@ -28,10 +31,30 @@ create_api_module(
...
+    import os
+    class SuppressOutput(object):
+        def write(self, txt):
+            pass  # Do nothing on write
+        def flush(self):
+            pass  # Do nothing on flush
+
+    sys.stdout = SuppressOutput()
+    )");
+    } catch (const std::exception& e) {
```
Comment (egelja) [inline linter/src/pywrapper/runtime.cpp:49]:
```suggestion
    } catch (...) {
```

File: linter/src/spawning/spawning.hpp [hunk 9]
```diff
@@ -1,15 +1,25 @@
...
-#include <unistd.h>
+#include <boost/asio.hpp>
+#include <boost/process.hpp>
 
-#include <string>
+namespace ba = boost::asio;
 
 namespace nutc {
 namespace spawning {
+class SpawnerManager {
```
Comment (egelja) [inline linter/src/spawning/spawning.hpp:12]:
```suggestion
class LintProcessSpawner {
```

File: linter/src/spawning/spawning.hpp [hunk 10]
```diff
@@ -1,15 +1,25 @@
...
 
-#include "logging.hpp"
+#include "lint_result.hpp"
 
-#include <unistd.h>
+#include <boost/asio.hpp>
+#include <boost/process.hpp>
 
-#include <string>
+namespace ba = boost::asio;
```
Comment (egelja) [inline linter/src/spawning/spawning.hpp:8]: this should go inside the `spawning` namespace

File: linter/src/spawning/spawning.cpp [hunk 11]
```diff
@@ -1,35 +1,94 @@
...
+    out_pipe << algo_code << std::flush;
+    out_pipe.pipe().close();
+
+    auto kill_timer = std::make_shared<ba::steady_timer>(io_service_);
+    kill_timer->expires_after(ba::chrono::seconds(LINT_AUTO_TIMEOUT_SECONDS));
+
+    lint_result res;
+
+    kill_timer->async_wait([child, in_pipe, &res](const auto& ec) {
+        if (!ec) {
```
Comment (egelja) [inline linter/src/spawning/spawning.cpp:36]:
```suggestion
        if (!ec) return;
```

you have the condition inverted.

image

File: linter/src/spawning/spawning.cpp [hunk 12]
```diff
@@ -1,35 +1,94 @@
...
+        }
+    });
+
+    auto buffer = std::make_shared<ba::streambuf>();
+    async_read_until(
+        *in_pipe,
+        *buffer,
+        '\n',
+        [buffer, kill_timer, child, in_pipe, &res](const auto& ec, auto) {
+            if (!ec) {
```
Comment (egelja) [inline linter/src/spawning/spawning.cpp:60]:
```suggestion
            if (!ec) return;
```

you have condition inverted agani.

**Review by egelja — CHANGES_REQUESTED** [review]

File: linter/src/crow/crow.cpp (hunk 1 above)
Comment (glasss13) [reply linter/src/crow/crow.cpp:64]: Meant to leave a TODO about this, currently 200 is expected by the rust server so I am just preserving past behavior.

File: linter/src/crow/crow.cpp (hunk 2 above)
Comment (glasss13) [reply linter/src/crow/crow.cpp:77]: the current design is single threaded and we don't have a reason to change that right now since the linter is not very performance critical

File: linter/src/spawning/spawning.cpp (hunk 11 above)
Comment (glasss13) [reply linter/src/spawning/spawning.cpp:36]: image
I believe its correct

File: linter/src/crow/crow.cpp [hunk 13]
```diff
@@ -43,83 +44,43 @@ get_server_thread()
...
+                    algo_id,
+                    fmt::format(
+                        "[linter] FAILURE - could not find algo {} for id {}\n",
+                        algo_id,
+                        uid
+                    )
+                );
+                crow::json::wvalue response = crow::json::wvalue({
+                    {"linting_status",
+                     static_cast<int>(client::LintingResultOption::UNKNOWN)}
```
Comment (stevenewald) [inline linter/src/crow/crow.cpp:60]: nit: I'd rather send a specific string vs an int representing an enum

File: linter/src/firebase/fetching.cpp [hunk 14]
```diff
@@ -185,44 +173,62 @@ get_algo_status(const std::string& uid, const std::string& algo_id)
...
+            res = curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "DELETE");
+            if (res != CURLE_OK) {
+                return std::nullopt;
+            }
         }
 
         res = curl_easy_perform(curl);
         if (res != CURLE_OK) {
-            log_e(firebase, "curl_easy_perform() failed: {}", curl_easy_strerror(res));
+            return std::nullopt;
```
Comment (stevenewald) [inline linter/src/firebase/fetching.cpp:223]: nit: we should still probably log this

File: linter/src/lint/lint.cpp [hunk 15]
```diff
@@ -1,58 +1,57 @@
...
+#include <string>
+
 namespace nutc {
 namespace lint {
-bool
-lint(
-    const std::string& uid, const std::string& algo_id, std::stringstream& output_stream
-)
+
+static bool
```
Comment (stevenewald) [inline linter/src/lint/lint.cpp:14]: nit: anonymous namespace > static

File: linter/src/lint/lint.cpp [hunk 16]
```diff
@@ -1,58 +1,57 @@
...
+lint_result
+lint(const std::string& algo_code)
+{
+    std::string out_message = "[linter] starting to lint algorithm\n";
+    bool ok = nutc::pywrapper::create_api_module(mock_market_func);
+    if (!ok) {
+        out_message += "[linter] failed to create API module\n";
+        return {false, out_message};
+    }
+    ok = nutc::pywrapper::supress_stdout();
```
Comment (stevenewald) [inline linter/src/lint/lint.cpp:29]: I'm not sure if we actually want to supress stdout in the linter - people may use it for debug

File: linter/src/spawning/spawning.cpp [hunk 17]
```diff
@@ -1,35 +1,108 @@
...
+                }
+
+                auto error = glz::read_json<nutc::lint::lint_result>(res, message);
+                if (error) {
+                    res = {
+                        false,
+                        "Internal server error. Reach out to nuft@u.northwesten.edu "
+                        "for support"
+                    };
+                };
```
Comment (stevenewald) [inline linter/src/spawning/spawning.cpp:95]: I'm gonna predict we're gonna run into issues somewhere with this down the line (I ran into a million when doing it for the exchange, but they're more immediately visible there). Just something to keep in mind, not actionable now

File: linter/src/spawning/spawning.hpp [hunk 18]
```diff
@@ -1,15 +1,30 @@
...
+    const std::filesystem::path& spawner_binary_path();
+
+    LintProcessManager() : io_service_() {}
+
+    ~LintProcessManager() { io_service_.stop(); }
+
+    nutc::lint::lint_result spawn_client(const std::string&);
+
+private:
+    ba::io_service io_service_;
```
Comment (stevenewald) [inline linter/src/spawning/spawning.hpp:26]: io_service is deprecated, use io_context

File: linter/src/spawning/spawning.hpp [hunk 19]
```diff
@@ -1,15 +1,30 @@
...
+namespace ba = boost::asio;
 
-}
+class LintProcessManager {
+public:
+    const std::filesystem::path& spawner_binary_path();
+
+    LintProcessManager() : io_service_() {}
+
+    ~LintProcessManager() { io_service_.stop(); }
```
Comment (stevenewald) [inline linter/src/spawning/spawning.hpp:21]: I may be misunderstanding this. io service/context blocks the thread it runs on - does it run on another thread? If so, why don't we stop that thread? If not, why would we need this here when it's blocking the main thread?

**Review by stevenewald — CHANGES_REQUESTED** [review]
Great work - few relatively small comments. Ignore the nits if you want, they're subjective

File: linter/src/spawning/spawning.cpp (hunk 17 above)
Comment (glasss13) [reply linter/src/spawning/spawning.cpp:95]: With telling people to email nutf@u.northwestern.edu?

File: linter/src/crow/crow.cpp (hunk 13 above)
Comment (glasss13) [reply linter/src/crow/crow.cpp:60]: Would have to change the rust server for this so I think I'd rather leave this for another PR
